cryptoconditions/condition.py

```diff
@@ -74,63 +94,98 @@ def from_uri(serialized_condition):
+        try:
+            cost = qs_dict['cost'][0]
+        except (KeyError, IndexError):
+            raise ParsingError(
+                'Invalid condition format: "cost" parameter or value missing.')
+
+        if not re.match(INTEGER_REGEX, cost):
+            raise ParsingError('No or invalid cost provided')
+
+        fingerprint = regex_match.group(1)
         condition = Condition()
-        condition.type_id = int(pieces[1], 16)
-        condition.bitmask = int(pieces[2], 16)
-        condition.hash = base64.urlsafe_b64decode(base64_add_padding(pieces[3]))
-        condition.max_fulfillment_length = int(pieces[4])
-
+        condition.type_id = condition_type['type_id']
+        condition._subtypes = set()
+        if condition_type['class'].TYPE_CATEGORY == 'compound':
+            condition._subtypes.update(qs_dict['subtypes'][0].split(','))
+        condition.hash = base64.urlsafe_b64decode(
+            base64_add_padding(fingerprint))
+        condition.cost = int(cost)
         return condition
 
+    # TODO make into a class method instead as it fits the purpose of an
+    # alternative way to instantiate the class
     @staticmethod
-    def from_binary(reader):
+    def from_binary(data):
         """
         Create a Condition object from a binary blob.
 
         This method will parse a stream of binary data and construct a
         corresponding Condition object.
 
         Args:
-            reader (Reader): Binary stream implementing the Reader interface
+            data (bytes): Binary data representing the condition.
         Returns:
             Condition: Resulting object
         """
-        reader = Reader.from_source(reader)
-
-        # Instantiate condition
-        condition = Condition()
-        condition.parse_binary(reader)
-
-        return condition
+        asn1_condtion_obj, residue = der_decode(data, asn1Spec=Asn1Condition())
+        asn1_condition_dict = nat_encode(asn1_condtion_obj)
+        return Condition.from_asn1_dict(asn1_condition_dict)
 
+    # TODO make into a class method instead as it fits the purpose of an
+    # alternative way to instantiate the class
     @staticmethod
-    def from_dict(data):
-        """
-        Create a Condition object from a dict.
-
-        Args:
-            data (dict): Dictionary containing the condition payload
-
-        Returns:
-            Condition: Resulting object
-        """
+    def from_asn1_dict(asn1_dict):
+        asn1_type, value = asn1_dict.popitem()
+        registered_type = TypeRegistry.find_by_asn1_type(asn1_type)
+        # Instantiate condition
         condition = Condition()
-        condition.parse_dict(data)
+        condition.type_id = registered_type['type_id']
+        condition.hash = value['fingerprint']
+        condition.cost = value['cost']
+        condition._subtypes = set()
+        if registered_type['class'].TYPE_CATEGORY == 'compound':
+            subtypes = {
```

Comment:
Nice use of itertools. But I feel that this could be in a separate method. Something like `subtypes_from_bitstring` and `subtypes_to_bitstring` for `to_ans1_dict`